os/pm/pm_metrics.c

```diff
@@ -17,117 +17,297 @@
-			timetoadd = time_diff(cur_time, ref_time);
-			isfirst = 0;
-		} else if (!nsc) {
-			/* Last node */
-			timetoadd = time_diff(cur_time, sc->timestamp);
-		} else if (ref_time >= sc->timestamp && (nsc->timestamp >= ref_time)) {
-			/* sc--reftime--nsc */
-			timetoadd = nsc->timestamp - ref_time;
-		} else {
-			timetoadd = nsc->timestamp - sc->timestamp;
-		}
+	clock_t now;
+	if (pm_metrics_running) {
+		now = clock_systimer();
+		g_pm_metrics->state_metrics.state_accum_ticks[g_pmglobals.state] += now - g_pm_metrics->state_metrics.stime;
+		g_pm_metrics->state_metrics.stime = now;
+	}
+}
+
+/****************************************************************************
+ * Name: pm_metrics_update_wakehandler
+ *
+ * Description:
+ *   This function is called inside pm_wakehandler. It counts the frequency of wakeup
+ *   sources, which are waking up the board. It also checks the amount of time board
+ *   was in sleep.
+ * 
+ * Input parameters:
+ *   missing_tick - the amount of time the board was in sleep.
+ *   wakeup_src   - the wakeup reason code.
+ *
+ * Returned value:
+ *   None
+ *
+ ****************************************************************************/
+void pm_metrics_update_wakehandler(clock_t missing_tick, pm_wakeup_reason_code_t wakeup_src)
+{
+	if (pm_metrics_running) {
+		g_pm_metrics->wakeup_src_counts[wakeup_src]++;
+		g_pm_metrics->board_sleep_ticks += missing_tick;
+	}
+}
 
-		s = sc->state;
-
-		pmvdbg(" Timetoadd is %d to state:%d\n", timetoadd, s);
-		switch (s) {
-		case PM_NORMAL:
-			normal_time += timetoadd;
-			break;
-		case PM_IDLE:
-			idle_time += timetoadd;
-			break;
-		case PM_STANDBY:
-			standby_time += timetoadd;
-			break;
-		case PM_SLEEP:
-			sleep_time += timetoadd;
-			break;
-		default:
-			printf("Invalid state\n");
+/****************************************************************************
+ * Name: pm_metrics
+ *
+ * Description:
+ *   This internal function is called to analyze the PM suspend and sleep behaviour.
+ *   It gathers the pm metrics statistics for provided time (in msec) to provide the
+ *   domain specific pm suspend information.
+ *
+ * Input Parameters:
+ *   milliseconds - the monitoring duration in milliseconds
+ *
+ * Returned Value:
+ *   OK (0)     - On Success
+ *   ERROR (-1) - On Error
+ *
+ ****************************************************************************/
+int pm_metrics(int milliseconds)
+{
+	clock_t start_time, end_time;
+	irqstate_t flags;
+	int index;
+	int n_domains;
+	/* If PM Metrics already running then notify other thread */
+	if (g_pm_metrics) {
+		pmdbg("PM Metrics already running\n");
+		return OK;
+	}
+	/* Lock PM so that no two thread can run PM Metrics simultaneously */
+	pm_lock();
+	/* Allocate memory for initializing PM Metrics measurements */
+	g_pm_metrics = pm_alloc(1, sizeof(pm_metric_t));
+	if (g_pm_metrics == NULL) {
+		set_errno(ENOMEM);
+		pmdbg("Unable to initialize pm_metrics, error = %d", get_errno());
+		return ERROR;
```

Comment:
Is it ok not to do pm_unlock here?

Reply:
Why we use pm_lock/pm_unlcok,
This seems calling in only here.

Reply:
pm_lock/pm_unlock is needed to make pm_metrics thread safe